File: GAS/src/fi/quanfoxes/Parser/patterns/UnarySignPattern.java
package fi.quanfoxes.Parser.patterns;

import fi.quanfoxes.Lexer.OperatorToken;
import fi.quanfoxes.Lexer.OperatorType;
import fi.quanfoxes.Lexer.Token;
import fi.quanfoxes.Lexer.TokenType;
import fi.quanfoxes.Parser.Node;
import fi.quanfoxes.Parser.Pattern;

import java.util.List;

public class UnarySignPattern extends Pattern {
    private static final int PRIORITY = 14;

    private static final int OPERATOR = 0;
    private static final int SIGN = 0;

    public UnarySignPattern() {
        super(TokenType.OPERATOR, TokenType.OPERATOR,
                TokenType.IDENTIFIER | TokenType.NUMBER | TokenType.PROCESSED);
    }

    @Override
    public int priority(List<Token> tokens) {
        return PRIORITY;
    }

    @Override
    public boolean passes(List<Token> tokens) {
        OperatorToken operator = (OperatorToken)tokens.get(OPERATOR);
        OperatorToken sign = (OperatorToken)tokens.get(SIGN);

        return operator.getOperator() != OperatorType.INCREMENT && operator.getOperator() != OperatorType.DECREMENT &&
                (sign.getOperator() == OperatorType.ADD || sign.getOperator() == OperatorType.SUBTRACT);
    }

    @Override
    public Node build(Node parent, List<Token> tokens) {
        return null;
    }
}

File: Vivid/Parser/Patterns/UnarySignPattern.cs
using System.Collections.Generic;

public class UnarySignPattern : Pattern
{
	public const int SIGN = 0;
	public const int OBJECT = 1;

	// Pattern 1: - $value
	// Pattern 2: + $value
	public UnarySignPattern() : base
	(
		TokenType.OPERATOR,
		TokenType.OBJECT
	)
	{ Priority = 18; }

	public override bool Passes(Context context, ParserState state, List<Token> tokens, int priority)
	{
		var sign = tokens[SIGN].To<OperatorToken>().Operator;
		if (sign != Operators.ADD && sign != Operators.SUBTRACT) return false;

		return state.Start == 0 || state.All[state.Start - 1].Is(TokenType.OPERATOR, TokenType.KEYWORD);
	}

	public override Node? Build(Context context, ParserState state, List<Token> tokens)
	{
		var value = Singleton.Parse(context, tokens[OBJECT]);
		var sign = tokens[SIGN].To<OperatorToken>().Operator;

		if (value.Instance == NodeType.NUMBER)
		{
			if (sign == Operators.SUBTRACT) value.To<NumberNode>().Negate();
			return value.To<NumberNode>();
		}

		return sign == Operators.SUBTRACT ? new NegateNode(value, tokens[SIGN].Position) : value;
	}
}